components/browser/session/src/main/java/mozilla/components/browser/session/engine/EngineObserver.kt

```diff
@@ -84,6 +86,13 @@ internal class EngineObserver(
             uri.query == originalUri.query
     }
 
+    private fun isSameSite(originalUrl: String, newUrl: String): Boolean {
```

Comment:
We also have a String extension function for this: `String.isSameOriginAs`, which should work here too?

Reply:
Awesome I didn't know that extension :)
I tried it but it throws an exception as soon as I load a website as it's not able to parse `about:blank`, additionally one of our [existing tests is failing](https://github.com/mozilla-mobile/android-components/blob/9120a9fde9fc5dc04d75df13ac04301829a20132/components/browser/session/src/test/java/mozilla/components/browser/session/engine/EngineObserverTest.kt#L49) with the same exception as at some point `session.url` is empty (I'm not sure if that could be possible in real life).

It looks like to be able to use it, we would have to do multiple validations before `String.isSameOriginAs` to avoid throwing in edge case. In that case, I would prefer the actual implementation. What do you think?  

```
java.net.MalformedURLException: no protocol: 

	at java.net.URL.<init>(URL.java:593)
	at java.net.URL.<init>(URL.java:490)
	at java.net.URL.<init>(URL.java:439)
	at mozilla.components.support.ktx.kotlin.StringKt$isSameOriginAs$1.invoke(String.kt:115)
	at mozilla.components.support.ktx.kotlin.StringKt.isSameOriginAs(String.kt:120)
	at mozilla.components.browser.session.engine.EngineObserver.onLocationChange(EngineObserver.kt:66)
	at mozilla.components.browser.session.engine.EngineObserverTest$engineSessionObserver$engineSession$1$loadUrl$1.invoke(EngineObserverTest.kt:84)
	at mozilla.components.browser.session.engine.EngineObserverTest$engineSessionObserver$engineSession$1$loadUrl$1.invoke(EngineObserverTest.kt:51)
	at mozilla.components.support.base.observer.ObserverRegistry.notifyObservers(ObserverRegistry.kt:137)
	at mozilla.components.concept.engine.EngineSession.notifyObservers(EngineSession.kt)
	at mozilla.components.browser.session.engine.EngineObserverTest$engineSessionObserver$engineSession$1.loadUrl(EngineObserverTest.kt:84)
	at mozilla.components.concept.engine.EngineSession.loadUrl$default(EngineSession.kt:426)
	at mozilla.components.browser.session.engine.EngineObserverTest.engineSessionObserver(EngineObserverTest.kt:92)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:601)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:260)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at androidx.test.ext.junit.runners.AndroidJUnit4.run(AndroidJUnit4.java:104)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)

```

Reply:
I would prefer fixing that case in the existing extension function. We can also move your `isSameSite` to a new ext. function too, but I don't like adding logic to `EngineObserver` as we want to get rid of `browser-session` and actively work toward refactoring everything.

Reply:
👍

Reply:
what do you think of this improvement to `String.isSameOriginAs`? When we get a `MalformedURLException` we report the urls are not the same.

```kotlin
fun String.isSameOriginAs(other: String): Boolean {
    fun canonicalizeOrigin(urlStr: String): String {
        val url = URL(urlStr)
        val port = if (url.port == -1) url.defaultPort else url.port
        val canonicalized = URL(url.protocol, url.host, port, "")
        return canonicalized.toString()
    }
    return try {
        canonicalizeOrigin(this) == canonicalizeOrigin(other)
    } catch (e: MalformedURLException) {
        false
    }
}
```

Reply:
@Amejia481 yes, I think that's fine for the purposes here. I would add it to the docs to say that false is returned if either this or `other` is not a valid URL...